src/crimson/os/seastore/segment_cleaner.h

```diff
@@ -44,6 +95,249 @@ struct segment_info_t {
+      }
+    }
+    iterator<is_const>& operator++() {
+      iter++;
+      auto& sm_info = segment_info_set->sm_info_vec[sm_id];
+      if (iter == sm_info->segment_infos.end()) {
+	device_id_t t_sm_id = sm_id;
+	while (++t_sm_id < max_devices && !segment_info_set->sm_info_vec[t_sm_id]);
+	if (t_sm_id < max_devices) {
+	  auto& sm = segment_info_set->sm_info_vec[t_sm_id];
+	  assert(sm);
+	  iter = sm->segment_infos.begin();
+	  sm_id = t_sm_id;
+	}
+      }
+      return *this;
+    }
+    bool operator==(iterator<is_const> rit) {
+      return sm_id == rit.sm_id && iter == rit.iter;
+    }
+    bool operator!=(iterator<is_const> rit) {
+      return !(*this == rit);
+    }
+    template <bool c = is_const, std::enable_if_t<c, int> = 0>
+    const SegmentInfoT& operator*() {
+      return *iter;
+    }
+    template <bool c = is_const, std::enable_if_t<!c, int> = 0>
+    SegmentInfoT& operator*() {
+      return *iter;
+    }
+    template <bool c = is_const, std::enable_if_t<c, int> = 0>
+    const SegmentInfoT* operator->() {
+      return &(*iter);
+    }
+    template <bool c = is_const, std::enable_if_t<!c, int> = 0>
+    SegmentInfoT* operator->() {
+      return &(*iter);
+    }
+  private:
+    device_id_t sm_id = 0;
+    std::conditional_t<
+      is_const,
+      const segment_info_set_t*,
+      segment_info_set_t*> segment_info_set = nullptr;
+    std::conditional_t<
+      is_const,
+      typename std::vector<SegmentInfoT>::const_iterator,
+      typename std::vector<SegmentInfoT>::iterator> iter;
+  };
+  SegmentInfoT& operator[](segment_id_t id) {
+    auto d_id = id.device_id();
+    auto& sm_info = sm_info_vec[d_id];
+    assert(sm_info);
+    assert(sm_info->device_id == d_id);
+    auto& segment_info = sm_info->segment_infos[id.internal_segment_id()];
+    assert(segment_info.segment == id);
+    return segment_info;
+  }
+  const SegmentInfoT& operator[](segment_id_t id) const {
+    auto d_id = id.device_id();
+    auto& sm_info = sm_info_vec.at(d_id);
+    assert(sm_info);
+    assert(sm_info->device_id == d_id);
+    auto& segment_info = sm_info->segment_infos[id.internal_segment_id()];
+    assert(segment_info.segment == id);
+    return segment_info;
+  }
+  std::optional<segment_manager_info_t<SegmentInfoT>>&
+  operator[](device_id_t id) {
+    auto& sm_info = sm_info_vec[id];
+    assert(sm_info && sm_info->device_id == id);
+    return sm_info;
+  }
+  const std::optional<segment_manager_info_t<SegmentInfoT>>&
+  operator[](device_id_t id) const {
+    auto& sm_info = sm_info_vec[id];
+    assert(sm_info && sm_info->device_id == id);
+    return sm_info;
+  }
+  void clear() {
+    sm_info_vec.clear();
+    total_segments = 0;
+    total_bytes = 0;
+    journal_segments = 0;
+  }
+  template <typename T = SegmentInfoT,
+	    std::enable_if_t<std::is_default_constructible<T>::value, int> = 0>
+  void resize(std::vector<SegmentManager*> sms) {
+    assert(sm_info_vec.empty());
```

Comment:
This rather implies that you cannot call resize() twice.  Let's rename this to initialize_segment_managers()?

Reply:
Do we actually need this at all?  Why not just call add_segment_manager directly?

Reply:
> Do we actually need this at all? Why not just call add_segment_manager directly?

yeah, will change that